File: Scripts/test_revisit.mjs
#!/usr/bin/env node
/**
 * Tests for the Revisit arithmetic — docs/revisit.js.
 *
 *     node Scripts/test_revisit.mjs
 *
 * buildRevisit is the whole point of that file being separate: it is pure, it
 * takes its clock as an argument, and it returns plain objects. So it is
 * testable without a DOM — loaded into a vm context stocked with the one thing
 * it borrows from the page (TOPIC_BY_ID, built from the real taxonomy), and run
 * over a synthetic library whose every row is placed relative to a fixed NOW.
 *
 * Same shape as Scripts/test_importers.mjs.
 */
import fs from "node:fs";
import path from "node:path";
import vm from "node:vm";
import { fileURLToPath } from "node:url";

const ROOT = path.resolve(path.dirname(fileURLToPath(import.meta.url)), "..");
const read = (p) => fs.readFileSync(p, "utf8");

/* ── Load docs/revisit.js the way the browser does ── */
const ctx = vm.createContext({ console });
vm.runInContext(read(path.join(ROOT, "docs", "taxonomy.js")), ctx, { filename: "taxonomy.js" });
// index.html's one line, so the names in this test are the real topic names.
vm.runInContext("const TOPIC_BY_ID = Object.fromEntries(TAXONOMY.map(t => [t.id, t]));", ctx, { filename: "TOPIC_BY_ID" });
vm.runInContext(read(path.join(ROOT, "docs", "revisit.js")), ctx, { filename: "revisit.js" });
const { buildRevisit, Revisit } = vm.runInContext("({ buildRevisit, Revisit })", ctx);

/* ── Tiny runner ── */
let failures = 0, checks = 0;
function ok(cond, message) {
  checks++;
  if (cond) return;
  failures++;
  console.log(`FAIL ${message}`);
}
function eq(actual, expected, message) {
  const a = JSON.stringify(actual), b = JSON.stringify(expected);
  ok(a === b, `${message}\n     expected ${b}\n     got      ${a}`);
}
function group(name, fn) { console.log(`\n${name}`); fn(); }

/* ── Fixture ──────────────────────────────────────────────────────────────
   A library placed around a fixed NOW. `d` is days ago, `o` is times opened. */
const DAY = 86_400_000;
const NOW = Date.parse("2026-09-03T12:00:00.000Z");
let seq = 0;
const bork = ({ d, topic = null, sub = null, tags = [], platform = "web", o = 0, openedD = null, deleted = false }) => {
  const id = `b${++seq}`;
  return {
    id, url: `https://example.com/${id}`, title: `Bork ${id}`, platform, kind: "article",
    category_id: topic, subcategory: sub, tags,
    saved_at: new Date(NOW - d * DAY).toISOString(),
    updated_at: new Date(NOW - d * DAY).toISOString(),
    deleted_at: deleted ? new Date(NOW).toISOString() : null,
    open_count: o,
    last_opened_at: o ? new Date(NOW - (openedD ?? 1) * DAY).toISOString() : null,
  };
};

const ROWS = [
  // ── This week (7): 3 Instagram, 2 X, 2 YouTube
  bork({ d: 0.5, topic: "fitness", platform: "instagram", o: 3, openedD: 0.2 }),
  bork({ d: 1, topic: "fitness", platform: "instagram", sub: "Mobility", tags: ["running"] }),
  bork({ d: 2, topic: "fitness", platform: "youtube", o: 9, openedD: 3 }),
  bork({ d: 3, topic: "recipes", platform: "instagram" }),
  bork({ d: 4, topic: "recipes", platform: "x" }),
  bork({ d: 5, topic: "travel", platform: "x", o: 3, openedD: 0.1 }),
  bork({ d: 6, topic: "history", platform: "youtube" }),
  // ── Last week (3): fitness had one, recipes none
  bork({ d: 8, topic: "fitness", platform: "x" }),
  bork({ d: 9, topic: "history", platform: "x", o: 1 }),
  bork({ d: 10, topic: "travel", platform: "web" }),
  // ── Older, inside 30 days
  bork({ d: 16, topic: "fitness", platform: "instagram" }),
  bork({ d: 20, topic: "recipes", platform: "instagram" }),
  bork({ d: 24, topic: "fitness", platform: "youtube" }),
  bork({ d: 27, topic: "travel", platform: "tiktok" }),
  // ── A month ago today (28–35): three of them, plus one just outside
  bork({ d: 29, topic: "art", platform: "x" }),
  bork({ d: 31, topic: "crypto", platform: "x" }),
  bork({ d: 34, topic: "crypto", platform: "youtube" }),
  bork({ d: 36, topic: "crypto", platform: "x" }),   // 36 days: outside "a month ago"
  // ── The 30 days before that
  bork({ d: 40, topic: "crypto", platform: "x" }),
  bork({ d: 44, topic: "investing", platform: "youtube" }),
  bork({ d: 48, topic: "investing", platform: "x" }),
  bork({ d: 52, topic: "history", platform: "web" }),
  // ── Never counted: a tombstone
  bork({ d: 12, topic: "fitness", platform: "x", deleted: true }),
];

const MISSIONS = [
  { id: "m1", title: "Train for running", category_id: "fitness", bookmark_ids: ["b1"],
    todos: [{ id: "t1", text: "Book a 10k", done: true }, { id: "t2", text: "Buy shoes", done: false }],
    is_archived: false, deleted_at: null, updated_at: "2026-09-01T00:00:00.000Z" },
  { id: "m2", title: "Learn pottery", category_id: "crafts", bookmark_ids: [],
    todos: [{ id: "t3", text: "Find a studio", done: true }],
    is_archived: false, deleted_at: null, updated_at: "2026-09-01T00:00:00.000Z" },   // nothing left
  { id: "m3", title: "Archived", category_id: null, bookmark_ids: [],
    todos: [{ id: "t4", text: "x", done: false }],
    is_archived: true, deleted_at: null, updated_at: "2026-09-01T00:00:00.000Z" },
  { id: "m4", title: "Deleted", category_id: null, bookmark_ids: [],
    todos: [{ id: "t5", text: "x", done: false }],
    is_archived: false, deleted_at: "2026-09-02T00:00:00.000Z", updated_at: "2026-09-02T00:00:00.000Z" },
];

const D = buildRevisit(ROWS, MISSIONS, NOW);

group("Shape", () => {
  eq(D.total, ROWS.length - 1, "tombstones are not part of the library");
  eq(D.thin, false, "22 borks is not a thin library");
  eq(Object.keys(D).sort(), ["monthAgo", "never", "quests", "repeat", "shifting", "thin", "total", "week"],
    "every section has a slot, empty or not");
});

group("1 · Saved this week", () => {
  eq(D.week.n, 7, "seven saved in the last seven days");
  eq(D.week.platforms, [{ key: "instagram", n: 3 }, { key: "x", n: 2 }, { key: "youtube", n: 2 }],
    "sources split, biggest first");
  eq(D.week.rising.map(t => [t.name, t.delta]), [["Fitness", 2], ["Recipes", 2]],
    "topics that grew on last week, biggest jump first — a topic that held steady is not growth");
  ok(D.week.rising.length <= 3, "at most three rising topics");
  ok(D.week.rising.every(t => t.id && t.name), "a rising topic carries the id goTopic needs, and its name");

  const quiet = buildRevisit(ROWS.filter(r => Date.parse(r.saved_at) < NOW - 7 * DAY), MISSIONS, NOW);
  eq(quiet.week, null, "a week with nothing saved has no section at all");
});

group("2 · You keep coming back to", () => {
  eq(D.repeat.rows.map(r => r.open_count), [9, 3, 3, 1], "most-opened first");
  ok(D.repeat.rows.length <= 5, "five at most");
  // b6 (travel) and b1 (fitness) are both on 3; b6 was opened more recently.
  eq(D.repeat.rows.slice(1, 3).map(r => r.id), ["b6", "b1"], "ties break on the most recent open");
  ok(D.repeat.rows.every(r => r.open_count > 0), "a bork you never opened is not one you come back to");

  const unopened = buildRevisit(ROWS.map(r => ({ ...r, open_count: 0, last_opened_at: null })), MISSIONS, NOW);
  eq(unopened.repeat, null, "no opens, no section");
});

group("3 · Saved, never opened", () => {
  ok(D.never.rows.every(r => !r.open_count), "every row is unopened");
  ok(D.never.rows.every(r => Date.parse(r.saved_at) < NOW - 7 * DAY), "nothing saved this week is called ignored");
  eq(D.never.n, 14, "fourteen saved and never looked at");
  eq(D.never.shown, 12, "twelve on screen");
  eq(D.never.more, 2, "and two more behind the button");
  const dates = D.never.rows.map(r => r.saved_at);
  eq([...dates].sort().reverse(), dates, "newest first");

  const small = buildRevisit(ROWS.slice(0, 12).map(r => ({ ...r, open_count: 0 })), MISSIONS, NOW);
  eq(small.never.more, 0, "under twelve, there is nothing more to show");
});

group("4 · A month ago today", () => {
  eq(D.monthAgo.rows.map(r => r.id), ["b15", "b16", "b17"], "28–35 days back, newest first");
  ok(D.monthAgo.rows.length <= 3, "three at most");
  const none = buildRevisit(ROWS.filter(r => Date.parse(r.saved_at) > NOW - 28 * DAY), MISSIONS, NOW);
  eq(none.monthAgo, null, "no month-old saves, no section");
});

group("5 · Side quests in progress", () => {
  eq(D.quests.missions.map(m => m.id), ["m1"], "only quests with a to-do left — never archived or deleted ones");
  eq(buildRevisit(ROWS, [], NOW).quests, null, "no quests, no section");
  eq(buildRevisit(ROWS, null, NOW).quests, null, "missing quests are not a crash");
});

group("6 · What's shifting", () => {
  eq(D.shifting.line, "Less Crypto, more Fitness, less Investing", "the sentence, biggest move first");
  eq(D.shifting.items.map(i => i.direction), ["down", "up", "down"], "each move carries which way it went");
  ok(!D.shifting.items.some(i => i.id === "history"), "a topic that barely moved is not a shift");
  ok(D.shifting.items.length <= 3, "three at most");
  ok(D.shifting.items[0].prevShare > D.shifting.items[0].share, "\"less\" means a smaller share than before");

  const thin = ROWS.slice(0, 12);
  eq(buildRevisit(thin, MISSIONS, NOW).shifting, null, "under 20 borks across both windows, we don't claim a trend");
  const onlyNew = ROWS.filter(r => Date.parse(r.saved_at) > NOW - 30 * DAY);
  eq(buildRevisit([...onlyNew, ...onlyNew.map((r, i) => ({ ...r, id: `dup${i}` }))], MISSIONS, NOW).shifting, null,
    "no 'before' window is not a trend either");
});

group("Thin library", () => {
  const few = buildRevisit(ROWS.slice(0, 9), MISSIONS, NOW);
  eq(few.total, 9, "nine borks");
  eq(few.thin, true, "under ten is thin");
  eq(buildRevisit(ROWS.slice(0, 10), MISSIONS, NOW).thin, false, "ten is not");
  eq(buildRevisit([], [], NOW).total, 0, "an empty library is a number, not an exception");
});

group("Purity", () => {
  const before = JSON.stringify(ROWS);
  buildRevisit(ROWS, MISSIONS, NOW);
  eq(JSON.stringify(ROWS), before, "buildRevisit never touches the rows it is given");
  const shuffled = [...ROWS].reverse();
  eq(JSON.stringify(buildRevisit(shuffled, MISSIONS, NOW)), JSON.stringify(D), "the answer does not depend on input order");
  eq(JSON.stringify(buildRevisit(ROWS, MISSIONS, NOW)), JSON.stringify(D), "same input, same answer");
  ok(Revisit.NEVER_SHOWN === 12, "twelve is the fold");
});

group("Junk in", () => {
  const messy = [null, undefined, { id: "x" }, { id: "y", saved_at: "not a date", open_count: "3" }];
  const d = buildRevisit(messy, [null, { id: "q" }], NOW);
  eq(d.total, 2, "rows that are not rows are dropped, the rest are counted");
  eq(d.never, null, "a bork with no saved date is not evidence of anything");
  eq(d.quests, null, "a quest with no to-dos is not in progress");
  eq(buildRevisit(null, null, null).total, 0, "no rows at all is 0, not a throw");
});

console.log(`\n${checks - failures}/${checks} checks passed`);
process.exit(failures ? 1 : 0);

File: Scripts/test_revisit.swift
import Foundation

/// Compile:
/// `swiftc -parse-as-library -o /tmp/revisit-tests Core/Platform.swift Core/Copy.swift Core/Revisit.swift Scripts/test_revisit.swift`

@main
enum RevisitTests {
    static func main() {
        var failures = 0
        func expect(_ cond: Bool, _ message: String) {
            if cond { print("ok   \(message)") }
            else { failures += 1; print("FAIL \(message)") }
        }

        var calendar = Calendar(identifier: .gregorian)
        calendar.timeZone = TimeZone(identifier: "UTC")!

        // A fixed "now" so nothing here depends on the day it runs.
        let now = calendar.date(from: DateComponents(year: 2026, month: 9, day: 3, hour: 14))!
        let today = calendar.startOfDay(for: now)
        func daysAgo(_ days: Int, hour: Int = 9) -> Date {
            calendar.date(byAdding: .hour, value: hour,
                          to: calendar.date(byAdding: .day, value: -days, to: today)!)!
        }

        var serial = 0
        func bork(
            _ days: Int,
            platform: Platform = .instagram,
            topic: (id: String, name: String)? = nil,
            opens: Int = 0,
            lastOpened: Date? = nil,
            title: String? = nil
        ) -> RevisitBork {
            serial += 1
            return RevisitBork(
                id: "b\(serial)",
                title: title ?? "bork \(serial)",
                platform: platform,
                topicID: topic?.id,
                topicName: topic?.name,
                savedAt: daysAgo(days),
                openCount: opens,
                lastOpenedAt: lastOpened
            )
        }

        let fitness = (id: "fitness", name: "Fitness")
        let crypto = (id: "crypto", name: "Crypto")
        let recipes = (id: "recipes", name: "Recipes")

        // MARK: Thin state

        let thin = Revisit.build(borks: (0..<9).map { bork($0) }, quests: [], now: now, calendar: calendar)
        expect(thin.isThin, "nine borks is still the thin state")
        expect(thin.total == 9, "the thin state still counts what is there")

        let notThin = Revisit.build(borks: (0..<10).map { bork($0) }, quests: [], now: now, calendar: calendar)
        expect(!notThin.isThin, "ten borks fills the screen in")

        // MARK: 1 — Saved this week

        let week = Revisit.build(
            borks: [
                bork(0, platform: .instagram, topic: fitness),
                bork(1, platform: .instagram, topic: fitness),
                bork(2, platform: .x, topic: fitness),
                bork(3, platform: .youtube, topic: recipes),
                // Last week: Fitness had one, Recipes had three.
                bork(8, platform: .x, topic: fitness),
                bork(9, platform: .x, topic: recipes),
                bork(10, platform: .x, topic: recipes),
                bork(11, platform: .x, topic: recipes),
                // Older than both windows.
                bork(40, platform: .web, topic: recipes),
            ],
            quests: [], now: now, calendar: calendar
        )
        let thisWeek = week.thisWeek
        expect(thisWeek?.count == 4, "this week counts the last seven days only")
        expect(
            thisWeek?.headline == "4 borks · 2 Instagram · 1 X · 1 YouTube",
            "the headline splits by platform, biggest first — got \(thisWeek?.headline ?? "nil")"
        )
        expect(
            thisWeek?.rising.first?.id == "fitness",
            "Fitness grew 3→ this week against 1 last week and leads"
        )
        expect(
            thisWeek?.rising.contains { $0.id == "recipes" } == false,
            "Recipes shrank and is not listed as rising"
        )
        expect(
            thisWeek?.rising.first?.previous == 1,
            "growth carries last week's number, not just this week's"
        )

        let quietWeek = Revisit.build(borks: [bork(20), bork(30)], quests: [], now: now, calendar: calendar)
        expect(quietWeek.thisWeek == nil, "a week with no saves hides the section")

        // MARK: 2 — You keep coming back to

        let opened = Revisit.build(
            borks: [
                bork(30, opens: 2, lastOpened: daysAgo(1)),
                bork(31, opens: 9, lastOpened: daysAgo(20)),
                bork(32, opens: 2, lastOpened: daysAgo(9)),
                bork(33, opens: 0),
                bork(34, opens: 1, lastOpened: daysAgo(2)),
                bork(35, opens: 5, lastOpened: daysAgo(30)),
                bork(36, opens: 4, lastOpened: daysAgo(30)),
                bork(37, opens: 3, lastOpened: daysAgo(30)),
            ],
            quests: [], now: now, calendar: calendar
        )
        expect(opened.comingBack.count == 5, "the strip is capped at five")
        expect(opened.comingBack.map(\.openCount) == [9, 5, 4, 3, 2], "sorted by open count")
        expect(
            opened.comingBack.last?.lastOpenedAt == daysAgo(1),
            "ties on open count break by most recently opened"
        )
        expect(
            opened.comingBack.allSatisfy { $0.openCount > 0 },
            "a never-opened bork is never something you keep coming back to"
        )

        // MARK: 3 — Saved, never opened

        let stale = Revisit.build(
            borks: [
                bork(2, opens: 0),                       // too new to count
                bork(8, opens: 0), bork(9, opens: 0),
                bork(30, opens: 0), bork(31, opens: 1),  // opened once — not stale
            ] + (0..<12).map { bork(20 + $0, opens: 0) },
            quests: [], now: now, calendar: calendar
        )
        expect(stale.neverOpened?.total == 15, "never-opened counts everything older than a week")
        expect(stale.neverOpened?.items.count == 15, "the model keeps all of them for the full list")
        expect(stale.neverOpened?.preview.count == Revisit.neverOpenedCap, "the section shows twelve")
        expect(stale.neverOpened?.overflow == 3, "the overflow is what the cap left out")
        expect(
            stale.neverOpened?.headline == "15 things you saved and never looked at.",
            "the copy names the number — got \(stale.neverOpened?.headline ?? "nil")"
        )
        expect(
            (stale.neverOpened?.items.first?.savedAt ?? .distantPast)
                > (stale.neverOpened?.items.last?.savedAt ?? .distantFuture),
            "newest first"
        )
        expect(
            stale.neverOpened?.preview.first?.id == stale.neverOpened?.items.first?.id,
            "the preview is the head of the same list"
        )

        let one = Revisit.build(borks: [bork(20, opens: 0)], quests: [], now: now, calendar: calendar)
        expect(one.neverOpened?.headline == "1 thing you saved and never looked at.", "singular reads properly")

        let allOpened = Revisit.build(borks: [bork(20, opens: 1, lastOpened: now)], quests: [], now: now, calendar: calendar)
        expect(allOpened.neverOpened == nil, "nothing stale hides the section")

        // MARK: 4 — A month ago

        let month = Revisit.build(
            borks: [
                bork(27), bork(28), bork(30), bork(33), bork(35), bork(36),
            ],
            quests: [], now: now, calendar: calendar
        )
        expect(month.monthAgo.count == 3, "capped at three")
        expect(
            month.monthAgo.allSatisfy { $0.savedAt <= daysAgo(28) && $0.savedAt >= daysAgo(35) },
            "only the 28–35 day window"
        )
        expect(
            month.monthAgo.first?.savedAt == daysAgo(28),
            "newest of the window first"
        )

        // MARK: 5 — Side quests in progress

        let quests = Revisit.build(
            borks: [bork(1)],
            quests: [
                RevisitQuest(id: "q1", title: "Improve mobility", remaining: 2),
                RevisitQuest(id: "q2", title: "Finished thing", remaining: 0),
            ],
            now: now, calendar: calendar
        )
        expect(quests.quests.map(\.id) == ["q1"], "only quests with steps left")

        // MARK: 6 — What's shifting

        // 12 recent + 12 earlier = 24, past the floor of 20. Fitness goes
        // 4/12 → 9/12; Crypto goes 8/12 → 3/12.
        let shifting = Revisit.build(
            borks: (0..<9).map { bork(2 + $0 % 20, topic: fitness) }
                + (0..<3).map { _ in bork(5, topic: crypto) }
                + (0..<4).map { _ in bork(40, topic: fitness) }
                + (0..<8).map { _ in bork(45, topic: crypto) },
            quests: [], now: now, calendar: calendar
        )
        expect(shifting.shifts.count == 2, "two topics moved — got \(shifting.shifts.count)")
        expect(
            shifting.shiftSentence == "More Fitness, less Crypto",
            "the sentence is the section — got \(shifting.shiftSentence ?? "nil")"
        )
        expect(shifting.shifts.first?.isUp == true, "the biggest mover leads")

        let tooFew = Revisit.build(
            borks: (0..<5).map { _ in bork(2, topic: fitness) } + (0..<5).map { _ in bork(45, topic: crypto) },
            quests: [], now: now, calendar: calendar
        )
        expect(tooFew.shifts.isEmpty, "under twenty borks across both windows says nothing")
        expect(tooFew.shiftSentence == nil, "and offers no sentence")

        let steady = Revisit.build(
            borks: (0..<12).map { _ in bork(2, topic: fitness) } + (0..<12).map { _ in bork(45, topic: fitness) },
            quests: [], now: now, calendar: calendar
        )
        expect(steady.shifts.isEmpty, "a library that did not change reports no shift")

        // MARK: Nothing at all

        let nothing = Revisit.build(borks: [], quests: [], now: now, calendar: calendar)
        expect(nothing.hasNothing, "an empty library has nothing in any section")
        expect(nothing.total == 0 && nothing.isThin, "and is thin")

        if failures > 0 { print("\n\(failures) failed"); exit(1) }
        print("\nall passed")
    }
}
